날씨앱 [STEP1] 미뇽

GuMinYoung opened the pull request:
@ryan-son 
안녕하세요! 1주차 코드리뷰 감사드리며, 2주차 PR 보냅니다.
처음에 브랜치 지정을 잘못한 것 같아서 새로 작성합니다. 내용은 동일합니다 (_ _)

## 고민했던 점
### SRP, ISP, DIP
- WeatherTableViewCell의 layViews에서 모든 UI 컴포넌트 초기화 및 레이아웃을 설정하고 있기 때문에 이를 컴포넌트별로 분리했습니다.
- fetchWeatherJSON 메서드가 fetch 기능만 수행하도록 네비게이션 타이틀을 설정하는 메서드를 별도로 구현해 분리했습니다.
- 화면을 그리는 역할은 View에서 수행하도록 DetailView를 생성해 WeatherDetailViewController에서 책임을 덜어냈습니다.
- 상세정보를 추상화해 WeatherDetailViewController, DetailView와의 의존성을 분리했습니다.
- 중복 코드를 제거하기 위해 싱글톤 패턴과 extension을 사용해서 이미지캐싱, DateFormatter 생성을 처리했습니다.
- 중복 코드를 제거하기 위해 타입프로퍼티로 tempUnit(화씨/섭씨) 값을 관리합니다. 처음에는 싱글톤을 떠올렸으나 값을 읽고 쓰는 단순한 작업만 필요하기 때문에 Shared 클래스에 타입프로퍼티만 두었습니다.
- UITableViewDataSource/Delegate를 각각 추상화해 ISP와 DIP를 실현해보았습니다.
### 기타
- ImageChacher의 load 함수에서 UIImage를 리턴할까 하였으나, 호출한 곳에서 UIImage를 받아 결국에는 UI 업데이트 작업을 수행할테니 메인스레드로 감싸 클로저를 사용하는 completion 형태로 작업했습니다.

 ---

## 해결이 되지 않은 점
### 의존성 주입 시점
- WeatherListViewController와 TableView DataSource/Delegate의 의존성을 분리했습니다. json 데이터가 세팅되면 WeatherListViewController의 setTableDataSourceAndDelegate 메서드에서 인스턴스를 생성해 의존성을 주입합니다. 그런데 1주차 코드리뷰에 의하면 이 방식은 유연성이 떨어지는 것 같아요. (이니셜라이저 내부 또는 메서드 내부에서 필요한 인스턴스를 생성하는 방식은 인스턴스로 변경해 사용할 수 없으므로 유연성이 떨어진다고 말씀 주셨습니다.) 이를 개선하기 위해서는 setTableDataSourceAndDelegate 매개변수로 WeatherListTableViewAdopter를 받아서 처리해면 되는 걸까요?
### 구체 타입 추상화
- WeatherListViewController에서 WeatherJSON 구체 타입의 프로퍼티를 가지고 있습니다. 다른 타입의 데이터가 들어와도 코드를 적게 수정하려면 이것도 추상화하면 좋을 것 같은데, 맞을까요?

---

## 조언을 얻고 싶은 부분
### VC에서 TableViewDataSource/Delegate를 분리했을 때의 이점
- ViewController에서 책임을 덜어냈으니 SRP 관점에서는 개선되었다고 생각합니다. 코드도 간결해졌구요. 그런데 DIP 관점에서는 잘 모르겠습니다. 특정 DataSource/Delegate에 구속 되지 않아서 좋은 걸까요? 현재 코드 기준으로는 WeatherListTableDataSource, WeatherListTableDataSource 내부에 있는 구체 타입([WeatherForecastInfo], City)까지 추상화해야 의존성이 완전리 분리될 것 같은데 제가 제대로 이해 했는지 궁금합니다. 이외의 장점도 함께 알려주시면 감사하겠습니다.
### willSet /  didSet 사용에 대한 고찰
- 이번 예제에서처럼, 현업에서도 특정 프로퍼티의 값이 변경된 후 수행하고 싶은 작업이 있을 때 willSet/didSet을 빈번하게 사용하고 있습니다. 프로젝트가 커질수록 그 내부에서 처리하는 코드도 많아지는데요. willSet/didSet에서 이렇게 온갖 작업을 다 해도 성능 이슈는 없는지, 다른 개선 방안이 있는지 질문 드립니다.
### 중첩함수와 SRP
- DetailView 클래스의 layViews메서드 한 곳에서 모든 UI 컴포넌트 초기화 및 레이아웃을 설정하고 싶지 않았습니다. 그런데 컴포넌트를 메서드 내부에서 선언했기 때문에 메서드 분리가 어려워 중첩함수로 사용했습니다. 이것도 SRP로 볼 수 있을까요?
### UI컴포넌트 선언
- 저는 UI컴포넌트를 클래스의 멤버변수로 두는 방식을 늘 사용해왔는데, 예제에서는 이렇게 하지 않고 메서드 내에서 선언해 사용하는 코드를 자주 접했습니다. 생각해보니 멤버변수로 둔다는 건 접근제어자를 무엇으로 지정했냐에 따라 여러 곳에서 접근할 수도 있다는 의미이고, 의존성 측면에서 코드 품질 저하가 일어날 우려가 있겠구나 싶었습니다. 뷰 내부에서 레이아웃 잡을 때만 쓴다면 예제처럼 메서드에서 선언하는 방식이 더 좋겠다는 생각이 들었습니다. 다만 생소한 방식이라.. 다른 분들도 이렇게 많이들 사용하는지,  사용한다면 제가 이해한 의도로 쓰는 게 맞는지, 또다른 이점이 있는지 궁금합니다.

GuMinYoung committed:
[Refactor] STEP1 - WeatherTableViewCell의 메서드 나누기 및 매직리터럴 제거

layViews에서 모든 UI 컴포넌트 초기화 및 레이아웃을 설정하고 있기 때문에 이를 분리한다.
reset 함수에서 사용하던 매직리터럴을 상수를 선언해 제거한다.

GuMinYoung committed:
[Refactor] STEP1 - ViewController의 네비게이션 타이틀 설정 메서드 분리

fetchWeatherJSON 메서드에서는 fetch 기능만 수행하도록 네비게이션 타이틀을 설정하는 메서드를 별도로 구현한다.
weatherJSON의 값이 바뀌면 네비게이션 타이틀을 설정하므로 willSet에서 네비게이션 타이틀을 설정 메서드를 호출한다.

Fixes: https://camp.yagom-academy.kr/camp/65cc2b983173745319e0464c/projects/65cc31bc3173745319e0465e

9min0-neowiz committed:
[Refactor] STEP1 - WeatherDetailViewController로부터 화면 그리는 역할을 분리

화면을 그리는 역할은 View에서 수행하도록 DetailView를 생성해 WeatherDetailViewController에서 책임을 덜어낸다.

Fixes: https://camp.yagom-academy.kr/camp/65cc2b983173745319e0464c/projects/65cc31bc3173745319e0465e

9min0-neowiz committed:
[Refactor] DetailView와 상세정보의 의존성 분리

DetailView가 구체 타입 WeatherForecastInfo, City, TempUnit 대신 추상화한 protocol에 의존하도록 한다.

Fixes: https://camp.yagom-academy.kr/camp/65cc2b983173745319e0464c/projects/65cc31bc3173745319e0465e

9min0-neowiz committed:
[Refactor] DetailViewController와 상세정보의 의존성 분리

DetailViewController가 구체 타입 WeatherForecastInfo, City, TempUnit 대신 추상화한 protocol에 의존하도록 한다.

Fixes: https://camp.yagom-academy.kr/camp/65cc2b983173745319e0464c/projects/65cc31bc3173745319e0465e

GuMinYoung committed:
[Refactor] 이미지캐셔 추가

URL로 이미지를 불러와 로드 및 캐시하는 코드가 중복되므로 이미지캐셔를 구현해 그 내부로 옮긴다.

Fixes: https://camp.yagom-academy.kr/camp/65cc2b983173745319e0464c/projects/65cc31bc3173745319e0465e

9min0-neowiz committed:
[Refactor] DateFormatter 중복 코드 제거 및 변수/클래스명 변경

DateFormatter를 여러 곳에서 생성하고 있으므로 Date의 extension으로 원하는 포맷, 스타일로 formatter를 만들게 하여 중복 코드를 제거한다.
타입을 따라 사용하던 변수명(infoProtocol)을 1주차 프로젝트 코드리뷰를 참고하여 변경한다.
ViewController라는 클래스명으로는 역할을 알 수 없기 때문에 WeatherListViewController로 변경한다.

9min0-neowiz committed:
[Refactor] UITableViewDataSource 분리

테이블뷰의 데이터를 제공하는 역할을 반드시 VC에서 담당할 필요는 없으므로 UITableViewDataSource를 별도 객체로 만든다.

Fixes: https://camp.yagom-academy.kr/camp/65cc2b983173745319e0464c/projects/65cc31bc3173745319e0465e
Ref: https://developer.apple.com/documentation/uikit/views_and_controls/table_views

9min0-neowiz committed:
[Refactor] UITableViewDelegate 분리

테이블뷰와 사용자 인터랙션 관리 역할을 VC에서 분리하기 위해 UITableViewDelegate를 별도 객체로 만든다.

Fixes: https://camp.yagom-academy.kr/camp/65cc2b983173745319e0464c/projects/65cc31bc3173745319e0465e
Ref: https://developer.apple.com/documentation/uikit/views_and_controls/table_views

9min0-neowiz committed:
[Refactor] 미사용 코드 제거 및 변수/클래스명 변경

리팩터링한 뒤 사용하지 않게 된 코드를 제거한다.
타입을 따라 사용하던 변수명(infoProtocol)을 1주차 프로젝트 코드리뷰를 참고하여 변경한다.
TableView에 쓰이는 객체이므로 클래스명을 WeatherListDataSource에서 WeatherListTableDataSource로 변경한다.

Fixes: https://camp.yagom-academy.kr/camp/65cc2b983173745319e0464c/projects/65cc31bc3173745319e0465e

9min0-neowiz committed:
[Refactor] DIP 적용 - TableView DataSource/Delegate 추상화

WeatherListViewController와 TableView DataSource/Delegate의 의존성을 분리하기 위해 프로토콜을 정의해 사용한다.

Fixes: https://camp.yagom-academy.kr/camp/65cc2b983173745319e0464c/projects/65cc31bc3173745319e0465e

9min0-neowiz committed:
[Refactor] 코드 정렬

줄바꿈 누락된 부분을 개선한다.

Fixes: https://camp.yagom-academy.kr/camp/65cc2b983173745319e0464c/projects/65cc31bc3173745319e0465e

9min0-neowiz committed:
[Refactor] tempUnit 중복 코드 제거

여러 클래스에서 보유한 tempUnit 프로퍼티는 모두 동일한 값을 공유하므로 타입프로퍼티를 사용해 접근하도록 하여 중복 코드를 제거한다.

Fixes: https://camp.yagom-academy.kr/camp/65cc2b983173745319e0464c/projects/65cc31bc3173745319e0465e

GuMinYoung committed:
[Feat/Refactor] STEP2 - 화씨 온도 표시 기능 추가 및 OCP 실현

기존 코드에서는 네비게이션바 우측 버튼을 클릭해도 섭씨/화씨 문구만 변경될 뿐 온도는 바뀌지 않는다.
수정한 코드에서는 섭씨/화씨에 따라 온도를 계산해 표시할 수 있다.
또한, 기존 코드에서는 switch문으로 섭씨/화씨 단위별 한글 문구, 기호를 설정하기 때문에 다른 단위가 추가될 때 switch문을 모두 찾아 수정해야 한다.
수정한 코드에서는 이를 방지하기 위해 switch문은 제거하고 열거형의 case로 관리하던 단위를 구조체로 만들어 타입(Metric, Imperial)으로 관리한다.
기호, 한글 문구는 공통으로 필요하므로 protocol로 정의하고 단위를 적용해 온도를 변환하는 applyTempUnit 메서드도 함께 추가한다.
이렇게 OCP를 실현하였고 새로운 단위가 추가될 경우 기존 코드 수정 없이 추가된 단위에 대한 코드만 작성하면 된다.

Fixes: https://camp.yagom-academy.kr/camp/65cc2b983173745319e0464c/projects/65cc31bc3173745319e0465e

ryan-son commented on WeatherForecast/WeatherForecast/Date+.swift:
```diff
@@ -0,0 +1,33 @@
+        formatter.timeStyle = style
+        return formatter.string(from: date)
+    }
+}
```
`DateFormatter`를 생성해 사용하는 것이 비용이 크다는 점을 들어보신 적이 있으신 것으로 보이네요. 생성보다는 설정해 string, date 메서드를 통해 반환받는 과정에서의 비용이 오히려 더 무시 못할 비용을 사용합니다. 특정 포매팅을 수행하는 DateFormatter 인스턴스를 타입 프로퍼티로 구성해서 사용해보는 전략도 생각해볼 수 있을 것 같습니다.

ryan-son commented on WeatherForecast/WeatherForecast/DetailView.swift:
```diff
@@ -0,0 +1,137 @@
+        self.weatherForecastInfo = weatherForecastInfo
+        self.cityInfo = cityInfo
+    }
+}
```
이 프로토콜로 인해 실제로 얻을 수 있는 이득이 있을까요? 데이터를 표현하는 역할을 가지는 모델 타입은 추가적인 추상화를 하지 않고 구조체 상태로 사용해도 괜찮을 것 같습니다.

ryan-son commented on WeatherForecast/WeatherForecast/DetailView.swift:
```diff
@@ -0,0 +1,137 @@
+        fatalError("init(coder:) has not been implemented")
+    }
+    
+    private func layViews() {
```
[Swift API Guidelines](https://www.swift.org/documentation/api-design-guidelines/) 작명 규칙에 따라 최대한 줄임 표현은 지양하는 것이 좋겠습니다.

<img width="841" alt="image" src="https://github.com/yagom-academy/refactoring-weather/assets/69730931/c860d06d-b4f5-40c6-ba19-8a09eddbdd2e">

ryan-son commented on WeatherForecast/WeatherForecast/DetailView.swift:
```diff
@@ -0,0 +1,137 @@
+                subview.numberOfLines = 1
+                subview.textAlignment = .center
+                subview.font = .preferredFont(forTextStyle: .body)
+            }
```
이 작업이 필요하다면 해당 속성을 설정해서 반환하는 convenience intializer 또는 커스텀 레이블을 정의해봐도 좋겠습니다.

ryan-son commented on WeatherForecast/WeatherForecast/DetailView.swift:
```diff
@@ -0,0 +1,137 @@
+        layLabels()
+        layMainStackView()
+        loadImage(to: iconImageView)
+    }
```
1. 한 칸 이상의 공백은 제거해주시는 것도 좋겠습니다.
2. 메서드 이름은 stackView 배치인데, 이미지 로드까지 수행하고 있네요. 이미지 로드 요청은 다른 곳에서 수행해보는 것이 좋겠습니다.

ryan-son commented on WeatherForecast/WeatherForecast/DetailView.swift:
```diff
@@ -0,0 +1,137 @@
+        ImageChacher.shared.load(urlString: urlString) { image in
+            imageView.image = image
+        }
+    }
```
이미지 캐시가 네트워킹을 통해 이미지 로드하는 역할까지 수행하고 있네요. 역할을 분리해서 아래와 같이 로직을 구성해보실 수 있을 것 같습니다.
```swift
if 캐시에 이미지가 있다면 {
    캐시 이미지 사용
} else {
    네트워킹을 통해 이미지 로드 후 캐시에 저장
}
```

ryan-son commented on WeatherForecast/WeatherForecast/ImageChacher.swift:
```diff
@@ -0,0 +1,44 @@
+
+import UIKit
+
+struct ImageChacher {
```
1. Cacher -> Cacher
2. `ImageCache`라는 이름도 좋을 것 같네요.

ryan-son commented on WeatherForecast/WeatherForecast/ImageChacher.swift:
```diff
@@ -0,0 +1,44 @@
+                completion(image)
+            }
+        }
+    }
```
1. 말씀드렸던 것처럼 이미지 캐셔는 이미지를 전달받아 캐시 공간에 저장하거나 저장된 이미지를 제공하는 역할만 수행하는 것이 좋겠습니다. 
2. 네트워킹 기능을 수행하는 `URLSession`은 항상 예상한 응답을 제공하지 않는 외부 의존성이기에 추상화와 의존성 분리를 해두시면 좋습니다.
3. Swift Concurrency를 이용한 로직을 작성하시고 계시니 로드 로직은 completionHandler 방식보다 async/await 방식으로 작성해보실 수 있으실 것 같습니다. async/await을 사용하면 completion을 호출하는 것을 잊을 일도 없을 겁니다. DispatchQueue.main.async도 지금까지 자주 사용해왔지만, Swift Concurrency를 이용하니 `@MainActor`, `MainActor.run { ... }`과 같은 API를 사용해보시는 것도 좋겠습니다.

ryan-son commented on WeatherForecast/WeatherForecast/TempUnit.swift:
```diff
@@ -0,0 +1,44 @@
+    func applyTempUnit(temp: Double) -> Double {
+        return temp * 1.8 + 32.0
+    }
+}
```
Metric, Imperial 단위를 추상화해보셨네요. 다른 온도 단위를 추가하기 좋겠어요.

ryan-son commented on WeatherForecast/WeatherForecast/Weather.swift:
```diff
@@ -48,15 +48,3 @@ class City: Decodable {
 }
-
-// MARK: - Temperature Unit
-enum TempUnit: String {
-    case metric, imperial
-    var expression: String {
-        switch self {
-        case .metric: return "℃"
-        case .imperial: return "℉"
-        }
-    }
-}
```
저는 개인적으로 두 단위 이외의 단위가 사용되지 않는다면 이런 형식의 구현도 괜찮다고 생각합니다.

ryan-son commented on WeatherForecast/WeatherForecast/WeatherDetailViewController.swift:
```diff
@@ -7,121 +7,27 @@
-        }
+        let weatherForecastInfo = info.weatherForecastInfo
+        let date: Date = Date(timeIntervalSince1970: weatherForecastInfo.dt)
+        navigationItem.title = Date.string(from: date, format: "yyyy-MM-dd(EEEEE) a HH:mm")
```
포맷을 문자열로 입력해주는 방식은 실수할 가능성이 높을 것 같네요. 포매팅해줄 포매터 인스턴스를 설정할 수 있거나 포매팅 문자열을 참조할 수 있는 방법을 마련해주는 것이 좋을 것 같습니다.

ryan-son commented on WeatherForecast/WeatherForecast/WeatherListTableDataSource.swift:
```diff
@@ -0,0 +1,51 @@
+
+import UIKit
+
+class WeatherListTableDataSource: NSObject, UITableViewDataSource {
```
상속하지 않아도 괜찮다면 final 키워드를 사용해볼 수 있을 것 같습니다.

ryan-son commented on WeatherForecast/WeatherForecast/WeatherListTableDataSource.swift:
```diff
@@ -0,0 +1,51 @@
+import UIKit
+
+class WeatherListTableDataSource: NSObject, UITableViewDataSource {
+    var weathers: [WeatherForecastInfo]
```
외부에서 참조하지 않아도 괜찮다면 private를 통해 접근 제한을 걸어줄 수 있을 것 같네요.

ryan-son commented on WeatherForecast/WeatherForecast/WeatherListTableDataSource.swift:
```diff
@@ -0,0 +1,51 @@
+        
+        guard let cell: WeatherTableViewCell = cell as? WeatherTableViewCell else {
+            return cell
+        }
```
프로토콜과 제네릭을 활용해 아래와 같은 메서드를 구성해볼 수 있을 것 같아요.

```swift
let cell: WeatherTableViewCell = tableView.dequeueReusableCell(for: indexPath)
```

위와 같은 구현이 가능하다면 등록도 아래 방식으로 가능하게 만들 수 있겠어요.

```swift
tableView.register(WeatherTableViewCell.self)
```

ryan-son commented on WeatherForecast/WeatherForecast/WeatherListTableDataSource.swift:
```diff
@@ -0,0 +1,51 @@
+        let weatherForecastInfo = weathers[indexPath.row]
+        cell.weatherLabel.text = weatherForecastInfo.weather.main
+        cell.descriptionLabel.text = weatherForecastInfo.weather.description
+        cell.temperatureLabel.text = "\(Shared.tempUnit.applyTempUnit(temp: weatherForecastInfo.main.temp))\(Shared.tempUnit.expression)"
```
표현 방식이 다소 길어지는데 개선해볼 수 있을까요?

ryan-son commented on WeatherForecast/WeatherForecast/WeatherListTableDataSource.swift:
```diff
@@ -0,0 +1,51 @@
+            if indexPath == tableView.indexPath(for: cell) {
+                cell.weatherIcon.image = image
+            }
+        })
```
이 로직을 View에서 분리해두면 유지보수 측면에서 개선될 것 같아요.

ryan-son commented on WeatherForecast/WeatherForecast/WeatherListViewController.swift:
```diff
@@ -0,0 +1,125 @@
+    var tableView: UITableView!
+    let refreshControl: UIRefreshControl = UIRefreshControl()
+    
+    var tableViewAdopter: WeatherListTableViewProtocol?
```
접근 제한을 설정해볼까요?

ryan-son commented on WeatherForecast/WeatherForecast/WeatherListViewController.swift:
```diff
@@ -0,0 +1,125 @@
+        tableViewAdopter = WeatherListTableViewAdopter(weatherListDataSource: weatherListDataSource, weatherListDelegate: weatherListDelegate)
+        tableView.dataSource = weatherListDataSource
+        tableView.delegate = weatherListDelegate
+    }
```
`weatherJSON`이 할당될 때마다 이 메서드가 호출되네요. 매번 새로운 `dataSource`와 `delegate`, `adoptor`를 생성해 할당해주기보다, `viewDidLoad` 시점에서 필요한 인스턴스를 생성해 할당해두고 데이터가 변경되면 `dataSource`에 새로운 데이터를 할당한 후 `reloadData`를 호출하는 방식으로 구성해보는 건 어떨까요?

ryan-son commented on WeatherForecast/WeatherForecast/WeatherListViewController.swift:
```diff
@@ -0,0 +1,125 @@
+}
+
+extension WeatherListViewController {
+    private func fetchWeatherJSON() {
+        let jsonDecoder: JSONDecoder = .init()
+        jsonDecoder.keyDecodingStrategy = .convertFromSnakeCase
+
+        guard let data = NSDataAsset(name: "weather")?.data else {
+            return
+        }
+        
+        let info: WeatherJSON
+        do {
+            info = try jsonDecoder.decode(WeatherJSON.self, from: data)
+        } catch {
+            print(error.localizedDescription)
+            return
+        }
+
+        weatherJSON = info
+    }
```
데이터 제공, 디코딩과 같은 역할을 분리해서 기능을 제공하는 타입들을 만들고, 해당 타입의 인스턴스에 View가 의존하도록 구성한다면 조금 더 구조를 유연하게 만들 수 있을 것 같아요.

ryan-son commented on WeatherForecast/WeatherForecast/WeatherTableViewCell.swift:
```diff
@@ -12,9 +12,11 @@ class WeatherTableViewCell: UITableViewCell {
     var temperatureLabel: UILabel!
     var weatherLabel: UILabel!
     var descriptionLabel: UILabel!
+    var dashLabel: UILabel!
```
Interface Builder 방식으로 View를 구성하는 것이 아니라면 암시적 옵셔널 타입은 사용하지 않는 것이 좋을 것 같아요. 옵셔널이 아닌 타입을 다루시고 싶으시다면 이곳에서 바로 인스턴스를 생성하는 방법도 사용해볼 수 있습니다.

```swift
private let dashLabel: UILabel = {
    let label = UILabel()
    label.textColor = .black
    label.font = .preferredFont(forTextStyle: .body)
    label.numberOfLines = 1
    return label
}()
// 또는 동일한 기능을 수행하는 convenience initializer 또는 커스텀 레이블 인스턴스 생성
```

GuMinYoung replied:
@ryan-son 
WeatherTableViewCell의 convenience initializer 말씀하시는 걸까요?
어떤 식으로 코드를 작성해야할지 잘 모르겠습니다. ^^;

GuMinYoung replied:
```swift
extension UILabel {
    convenience init(textColor: UIColor, font: UIFont, numberOfLines: Int) {
        self.init(frame: .zero)
        self.textColor = textColor
        self.font = font
        self.numberOfLines = numberOfLines
    }
}
```

혹시 이것일까용?

ryan-son commented on WeatherForecast/WeatherForecast/WeatherTableViewCell.swift:
```diff
@@ -93,11 +115,24 @@ class WeatherTableViewCell: UITableViewCell {
     }
 }
+
+enum Placeholder {
+    static let weatherIcon = "arrow.down.circle.dotted"
+    static let dateLabel = "0000-00-00 00:00:00"
+    static let temperatureLabel = "00℃"
+    static let weatherLabel = "~~~"
+    static let descriptionLabel = "~~~~~"
+}
```
인스턴스 생성이 필요하지 않아 열거형을 사용하셨네요. 좋습니다!

ryan-son reviewed (changes requested):
미뇽, 두번째 프로젝트는 작업 범위가 넓었을텐데 모두 수정해보시느라 고생 많으셨습니다.

이전과 마찬가지로 본문에서는 질문해주신 내용만 다루고, 세부적인 내용은 코멘트에서 다루도록 하겠습니다.

>## 해결이 되지 않은 점
>### 의존성 주입 시점
>- WeatherListViewController와 TableView DataSource/Delegate의 의존성을 분리했습니다. json 데이터가 세팅되면 WeatherListViewController의 setTableDataSourceAndDelegate 메서드에서 인스턴스를 생성해 의존성을 주입합니다. 그런데 1주차 코드리뷰에 의하면 이 방식은 유연성이 떨어지는 것 같아요. (이니셜라이저 내부 또는 메서드 내부에서 필요한 인스턴스를 생성하는 방식은 인스턴스로 변경해 사용할 수 없으므로 유연성이 떨어진다고 말씀 주셨습니다.) 이를 개선하기 위해서는 setTableDataSourceAndDelegate 매개변수로 WeatherListTableViewAdopter를 받아서 처리해면 되는 걸까요?

데이터를 다루거나 UITableViewDelegate가 수행하는 작업에 대한 관심사 분리를 위해 미뇽이 하신 방식으로 작업하시는 것도 좋은 방법입니다. 하지만 UITableView의 DataSource와 Delegate를 상황에 따라 변경해가며 사용하는 케이스는 드뭅니다. ViewController에 종속되는 경우가 다수죠. 그러니 DataSource와 Delegate를 외부로부터 주입받아 사용하지는 않아도 괜찮습니다. 필요치 않은 유연성은 오버엔지니어링이며 작업 시간만 늘어날 수 있습니다.

>### 구체 타입 추상화
>- WeatherListViewController에서 WeatherJSON 구체 타입의 프로퍼티를 가지고 있습니다. 다른 타입의 데이터가 들어와도 코드를 적게 수정하려면 이것도 추상화하면 좋을 것 같은데, 맞을까요?

ViewController에서 다루는 데이터는 상태가 관리되고, View에서 소비될 것이기 때문에 추상적일 필요가 없습니다. 오히려 타입의 데이터가 들어올 가능성이 생기면 유연함은 생겨도 데이터를 다루기가 더 까다로워질 수 있습니다.

---

>## 조언을 얻고 싶은 부분
>### VC에서 TableViewDataSource/Delegate를 분리했을 때의 이점
>- ViewController에서 책임을 덜어냈으니 SRP 관점에서는 개선되었다고 생각합니다. 코드도 간결해졌구요. 그런데 DIP 관점에서는 잘 모르겠습니다. 특정 DataSource/Delegate에 구속 되지 않아서 좋은 걸까요? 현재 코드 기준으로는 WeatherListTableDataSource, WeatherListTableDataSource 내부에 있는 구체 타입([WeatherForecastInfo], City)까지 추상화해야 의존성이 완전리 분리될 것 같은데 제가 제대로 이해 했는지 궁금합니다. 이외의 장점도 함께 알려주시면 감사하겠습니다.

SRP 관점의 개선이 있으며, 데이터를 다루거나 표시할 때, 그리고 UITableView와 상호작용할 때와 같은 작업을 관심사별로 분리하여 별도 타입에서 다루고 있기 때문에 로직의 수정이 필요할 때 영향을 받는 범위가 해당 타입으로 한정되어 유지보수하기가 용이해집니다. 위에서 말씀드린 것과 같은 맥락으로, ViewController별로 요구하는 UITableView의 형상이나 상호작용 방법, 그리고 데이터 표현 방식이 달라지므로 별도로 타입을 구성하는 것이 합리적입니다. 따라서 데이터를 표현하는 모델의 추상화까지는 필요하지 않다고 생각합니다.

>### willSet / didSet 사용에 대한 고찰
>- 이번 예제에서처럼, 현업에서도 특정 프로퍼티의 값이 변경된 후 수행하고 싶은 작업이 있을 때 willSet/didSet을 빈번하게 사용하고 있습니다. 프로젝트가 커질수록 그 내부에서 처리하는 코드도 많아지는데요. willSet/didSet에서 이렇게 온갖 작업을 다 해도 성능 이슈는 없는지, 다른 개선 방안이 있는지 질문 드립니다.

특별히 성능 이슈는 없습니다. 하지만 프로퍼티 옵저버에서 수행하는 로직이 많을수록 코드 흐름을 파악하기 어려워지는 건 사실이죠. 이런 점은 Closure 주입, NotificationCenter, KVO, Combine/RxSwift와 같은 반응형 프로그래밍 프레임워크를 사용하여 표현 방식을 다르게 변경할 수 있습니다.

>### 중첩함수와 SRP
>- DetailView 클래스의 layViews메서드 한 곳에서 모든 UI 컴포넌트 초기화 및 레이아웃을 설정하고 싶지 않았습니다. 그런데 컴포넌트를 메서드 내부에서 선언했기 때문에 메서드 분리가 어려워 중첩함수로 사용했습니다. 이것도 SRP로 볼 수 있을까요?

SRP 측면의 개선은 보통 타입의 분리로 일어난다고 할 수 있습니다. 여기에서는 메서드를 통해 수행 단위에 의미를 부여했으니 주석을 개선한 정도라고 볼 수 있을 것 같습니다. 의미 단위별로 코드를 파악하기 쉽도록 메서드를 분리하는 것은 좋습니다.

>### UI컴포넌트 선언

>- 저는 UI컴포넌트를 클래스의 멤버변수로 두는 방식을 늘 사용해왔는데, 예제에서는 이렇게 하지 않고 메서드 내에서 선언해 사용하는 코드를 자주 접했습니다. 생각해보니 멤버변수로 둔다는 건 접근제어자를 무엇으로 지정했냐에 따라 여러 곳에서 접근할 수도 있다는 의미이고, 의존성 측면에서 코드 품질 저하가 일어날 우려가 있겠구나 싶었습니다. 뷰 내부에서 레이아웃 잡을 때만 쓴다면 예제처럼 메서드에서 선언하는 방식이 더 좋겠다는 생각이 들었습니다. 다만 생소한 방식이라.. 다른 분들도 이렇게 많이들 사용하는지, 사용한다면 제가 이해한 의도로 쓰는 게 맞는지, 또다른 이점이 있는지 궁금합니다.

저도 멤버변수로 자주 선언합니다. 인스턴스 메서드에서 접근할 필요가 있다면 인스턴스 프로퍼티로 정의하는 것이 합리적이겠죠. 하지만 인스턴스 메서드 어느 곳에서도 해당 View 요소를 참조할 일이 없다면 메서드 내부에서 인스턴스를 생성해 사용하는 방법도 합리적이라고 할 수 있습니다. View 요소를 인스턴스 프로퍼티로 정의한다고 해서 코드 품질 저하를 우려하실 필요는 없다고 보입니다.

프로젝트 수행 중 막히는 부분이 있으시거나 리뷰 중 이해가 되지 않는 부분이 있으시면 언제든지 말씀해주세요! 고생 많으셨습니다.

9min0-neowiz committed:
[Refactor] STEP1 코드리뷰 반영

2주차 프로젝트 STEP1 코드리뷰를 반영해 개선한다.
폴더를 생성해 연관 소스파일끼리 관리한다.

Fixes: Fixes: https://camp.yagom-academy.kr/camp/65cc2b983173745319e0464c/projects/65cc31bc3173745319e0465e
Ref: https://github.com/yagom-academy/refactoring-weather/pull/32

9min0-neowiz committed:
Merge branch 'rft_2_refactoring2024' into temp

# Conflicts:
#	WeatherForecast/WeatherForecast.xcodeproj/project.pbxproj

9min0-neowiz committed:
[Rename] 소스파일 이동

브랜치 머지 과정에서 누락된 소스파일을 폴더에 넣는다.

9min0-neowiz committed:
[Refactor] 런타임 성능, 컴파일 타임 성능 최적화

동적 디스패치 보다는 정적 디스패치를 지향하기 위해, 상속/참조/identity에 따른 제어/Objective-C와의 상호운용이 필요 없는 class는 final class 또는 struct로 변경한다.
외부로 드러날 필요가 없는 프로퍼티와 메서드는 private으로 은닉화하여 캡슐화를 지향한다.

Fixes: https://camp.yagom-academy.kr/camp/65cc2b983173745319e0464c/projects/65cc31bc3173745319e0465e
Ref: 1주차 코드리뷰 https://github.com/yagom-academy/refactoring-parcel-invoice/pull/48,
클래스 vs 구조체 https://developer.apple.com/documentation/swift/choosing-between-structures-and-classes#Use-Classes-When-You-Need-Objective-C-Interoperability
